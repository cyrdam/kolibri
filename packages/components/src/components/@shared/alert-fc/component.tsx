import { h, type FunctionalComponent as FC } from '@stencil/core';
import type { JSXBase } from '@stencil/core/internal';
import clsx from 'clsx';

import { KolButtonWcTag, KolHeadingWcTag, KolIconTag } from '../../../core/component-names';
import type { AlertType, InternalAlertProps } from '../../../schema';
import { Log } from '../../../schema';
import { translate } from '../../../i18n';

const Icon: FC<{ ariaLabel: string; icon: string; label?: string }> = (props) => {
	return <KolIconTag class="heading-icon" _label={props.label ? '' : props.ariaLabel} _icons={props.icon} />;
};

const AlertIcon: FC<{ label?: string; type?: AlertType }> = ({ type, label }) => {
	switch (type) {
		case 'error':
			return <Icon ariaLabel={translate('kol-error')} icon="codicon codicon-error" label={label} />;
		case 'info':
			return <Icon ariaLabel={translate('kol-info')} icon="codicon codicon-info" label={label} />;
		case 'warning':
			return <Icon ariaLabel={translate('kol-warning')} icon="codicon codicon-warning" label={label} />;
		case 'success':
			return <Icon ariaLabel={translate('kol-success')} icon="codicon codicon-pass" label={label} />;
		default:
			return <Icon ariaLabel={translate('kol-message')} icon="codicon codicon-comment" label={label} />;
	}
};

export type KolAlertFcProps = JSXBase.HTMLAttributes<HTMLDivElement> &
	Partial<Omit<InternalAlertProps, 'on'>> & {
		onCloserClick?: () => void;
		onVibrationTimeout?: () => void;
	};

const KolAlertFc: FC<KolAlertFcProps> = (props, children) => {
	const { class: classNames = {}, type = 'default', variant = 'msg', label, hasCloser, alert, onVibrationTimeout, onCloserClick, level, ...other } = props;

	if (alert && navigator.vibrate) {
		/**
		 * - https://developer.mozilla.org/de/docs/Web/API/Navigator/vibrate
		 * - https://googlechrome.github.io/samples/vibration/
		 */
		try {
			Log.debug(['Navigator should vibrate ...', navigator.vibrate([100, 75, 100, 75, 100])]);
		} catch (e) {
			Log.debug('Navigator does not support vibration.');
		}

		setTimeout(() => {
			onVibrationTimeout?.();
		}, 10000);
	}

	const rootProps = {
		class: clsx('kol-alert-wc', 'alert', type, variant, { hasCloser: !!hasCloser }, classNames),
		role: alert ? 'alert' : undefined,
		...other,
	} as Partial<JSXBase.HTMLAttributes<HTMLDivElement>>;

	return (
		<div {...rootProps}>
			<div class="heading">
				<AlertIcon label={label} type={type} />
				<div class="heading-content">
					{label ? <KolHeadingWcTag _label={label} _level={level} /> : null}
					{variant === 'msg' && <div class="content">{children}</div>}
				</div>
				{hasCloser && (
					<KolButtonWcTag
						class="close"
						_ariaDescription={label?.trim() || ''}
						_hideLabel
						_icons={{
							left: {
								icon: 'codicon codicon-close',
							},
						}}
						_label={translate('kol-close-alert')}
						_on={{ onClick: onCloserClick }}
						_tooltipAlign="left"
					></KolButtonWcTag>
				)}
			</div>
			{variant === 'card' && <div class="content">{children}</div>}
		</div>
	);
};

export default KolAlertFc;
