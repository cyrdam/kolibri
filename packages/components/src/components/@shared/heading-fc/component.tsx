import clsx from 'clsx';
import type { FunctionalComponent as FC, JSX, VNode } from '@stencil/core';
import { h } from '@stencil/core';
import type { HeadingProps, HeadingVariantPropType } from '../../../schema';

function renderHeadline(props: { label?: string; variant?: HeadingVariantPropType; level?: number; children?: VNode[] }) {
	const validHeadline = typeof props.level === 'number' && props.level > 0 && props.level <= 6;
	const HeadlineTag = validHeadline ? `h${props.level}` : 'strong';
	const variant = props.variant || HeadlineTag;

	return (
		<HeadlineTag
			class={clsx({
				headline: true,
				[`headline-${variant}`]: true,
			})}
		>
			{props.label}
			{props.children}
		</HeadlineTag>
	);
}

function renderSecondaryHeadline(headline?: string, level?: number): JSX.Element | null {
	if (!headline) {
		return null;
	}

	switch (level) {
		case 1: // if the headline is only strong
			return <span class="secondary-headline">{headline}</span>;
		case 2: // if the headline is h1
			return <h2 class="secondary-headline">{headline}</h2>;
		case 3: // if the headline is h2
			return <h3 class="secondary-headline">{headline}</h3>;
		case 4: // if the headline is h3
			return <h4 class="secondary-headline">{headline}</h4>;
		case 5: // if the headline is h4
			return <h5 class="secondary-headline">{headline}</h5>;
		case 6: // if the headline is h5
			return <h6 class="secondary-headline">{headline}</h6>;
		default: // if the headline is h6
			return <strong class="secondary-headline">{headline}</strong>;
	}
}

const KolHeadingFc: FC<HeadingProps> = ({ _secondaryHeadline, _label, _level = 1 }, children) => {
	return (
		<section class="kol-heading-wc" style={{ display: 'inline' }}>
			{_secondaryHeadline ? (
				<hgroup>
					{renderHeadline({ label: _label, level: _level, children })}
					{renderSecondaryHeadline(_secondaryHeadline, _level + 1)}
				</hgroup>
			) : (
				renderHeadline({ label: _label, level: _level, children })
			)}
		</section>
	);
};

export default KolHeadingFc;
