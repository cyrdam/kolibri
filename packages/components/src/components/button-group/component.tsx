import type { ButtonGroupAPI, ButtonGroupStates } from '../../schema';
import type { JSX } from '@stencil/core';
import { Component, h, Host, State } from '@stencil/core';

/**
 * @deprecated This component has been deprecated as it does not contribute significantly to accessibility.
 * @internal
 */
@Component({
	tag: 'kol-button-group-wc',
	shadow: false,
})
export class KolButtonGroupWc implements ButtonGroupAPI {
	public render(): JSX.Element {
		return (
			<Host class="kol-button-group-wc">
				<slot />
			</Host>
		);
	}

	@State() public state: ButtonGroupStates = {};
}
