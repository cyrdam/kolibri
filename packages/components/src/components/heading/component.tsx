import type { HeadingAPI, HeadingLevel, HeadingStates, HeadingVariantPropType, LabelWithExpertSlotPropType } from '../../schema';
import { validateHeadingVariant, validateLabelWithExpertSlot, watchString } from '../../schema';
import { Component, h, Host, Prop, State, Watch } from '@stencil/core';

import { watchHeadingLevel } from './validation';

import type { JSX } from '@stencil/core';
import { KolHeadingFc } from '../@shared';

/**
 * @internal
 * @slot - Inhalt der Überschrift.
 */
@Component({
	tag: 'kol-heading-wc',
	shadow: false,
})
export class KolHeadingWc implements HeadingAPI {
	/**
	 * Defines the visible or semantic label of the component (e.g. aria-label, label, headline, caption, summary, etc.). Set to `false` to enable the expert slot.
	 */
	@Prop() public _label!: LabelWithExpertSlotPropType;

	/**
	 * Defines which H-level from 1-6 the heading has. 0 specifies no heading and is shown as bold text.
	 */
	@Prop() public _level?: HeadingLevel = 1;

	/**
	 * Setzt den Text einer weiteren Überschrift, einen Level kleiner, unter der Ersten.
	 */
	@Prop() public _secondaryHeadline?: string;

	/**
	 * Defines which variant should be used for presentation.
	 */
	@Prop() public _variant?: HeadingVariantPropType;

	@State() public state: HeadingStates = {
		_label: '', // ⚠ required
		_level: 1,
	};

	@Watch('_label')
	public validateLabel(value?: LabelWithExpertSlotPropType): void {
		validateLabelWithExpertSlot(this, value);
	}

	@Watch('_level')
	public validateLevel(value?: HeadingLevel): void {
		watchHeadingLevel(this, value);
	}

	@Watch('_secondaryHeadline')
	public validateSecondaryHeadline(value?: string): void {
		watchString(this, '_secondaryHeadline', value);
	}

	@Watch('_variant')
	public validateVariant(value?: HeadingVariantPropType): void {
		validateHeadingVariant(this, value);
	}

	public componentWillLoad(): void {
		this.validateLabel(this._label);
		this.validateLevel(this._level);
		this.validateSecondaryHeadline(this._secondaryHeadline);
		this.validateVariant(this._variant);
	}

	public render(): JSX.Element {
		return (
			<Host class="kol-heading-wc">
				<KolHeadingFc _label={this.state._label} _level={this.state._level} _secondaryHeadline={this.state._secondaryHeadline} _variant={this.state._variant}>
					<slot name="expert" />
				</KolHeadingFc>
			</Host>
		);
	}
}
