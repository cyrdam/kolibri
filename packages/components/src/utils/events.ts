// TODO: Should be synchronized with enums/events.ts
export type KoliBriEventType =
	| 'blur'
	| 'change'
	| 'changePage'
	| 'changePageSize'
	| 'click'
	| 'close'
	| 'focus'
	| 'input'
	| 'mousedown'
	| 'reset'
	| 'selection-change' // todo remove
	| 'selectionChange'
	| 'sort'
	| 'submit'
	| 'toggle';

export function stopPropagation(event: Event): void {
	event.stopImmediatePropagation();
	event.stopPropagation();
}

function createKoliBriEvent<T>(type: KoliBriEventType, detail: T): CustomEvent {
	const event = new CustomEvent(`kol-${type}`, {
		bubbles: true,
		cancelable: true,
		composed: true,
		detail: detail,
	});
	return event;
}

function dispatchKoliBriEvent<T>(target: EventTarget, type: KoliBriEventType, detail?: T): boolean {
	const dispatch = target.dispatchEvent(createKoliBriEvent(type, detail));
	return dispatch;
}

/**
 * @deprecated
 */
export function tryToDispatchKoliBriEvent<T>(type: KoliBriEventType, target?: EventTarget, detail?: T): void {
	target && dispatchKoliBriEvent(target, type, detail);
}

export function dispatchDomEvent(target: HTMLElement, type: KoliBriEventType) {
	target.dispatchEvent(new Event(type, { bubbles: true, composed: true }));
}

export function preventDefaultAndStopPropagation(event: Event) {
	event.preventDefault();
	stopPropagation(event);
}
