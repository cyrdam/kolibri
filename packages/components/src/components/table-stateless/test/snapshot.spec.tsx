import { KolTableStatelessWcTag } from '../../../core/component-names';
import type { TableStatelessProps } from '../../../schema';
import { executeSnapshotTests } from '../../../utils/testing';

import { KolTableStateless } from '../component';

executeSnapshotTests<TableStatelessProps>(
	KolTableStatelessWcTag,
	[KolTableStateless],
	[
		{
			_label: 'Table with horizontal and vertical headers',
			_headerCells: {
				horizontal: [
					[
						{ key: 'header1', label: 'Header1', textAlign: 'left' },
						{ key: 'header2', label: 'Header2', textAlign: 'center' },
						{ key: 'header3', label: 'Header3', textAlign: 'right' },
					],
				],
				vertical: [
					[
						{ key: 'row1', label: 'Row 1', textAlign: 'left' },
						{ key: 'row2', label: 'Row 2', textAlign: 'center' },
						{ key: 'row3', label: 'Row 3', textAlign: 'right' },
					],
				],
			},
			_data: [
				{ header1: 'Cell 1.1', header2: 'Cell 1.2', header3: 'Cell 1.3' },
				{ header1: 'Cell 2.1', header2: 'Cell 2.2', header3: 'Cell 2.3' },
				{ header1: 'Cell 3.1', header2: 'Cell 3.2', header3: 'Cell 3.3' },
			],
		},
		{
			_label: 'Table with only horizontal headers',
			_headerCells: {
				horizontal: [
					[
						{ key: 'header1', label: 'Header 1', textAlign: 'left' },
						{ key: 'header2', label: 'Header 2', textAlign: 'center' },
					],
				],
				vertical: [],
			},
			_data: [
				{ header1: 'Cell 1.1', header2: 'Cell 1.2' },
				{ header1: 'Cell 2.1', header2: 'Cell 2.2' },
			],
		},
		{
			_label: 'Table with two horizontal header rows',
			_headerCells: {
				horizontal: [
					[
						{ label: 'Header 1', textAlign: 'left' },
						{ label: 'Header 2', textAlign: 'center' },
					],
					[
						{ key: 'header1', label: 'Sub Header 1', textAlign: 'left' },
						{ key: 'header2', label: 'Sub Header 2', textAlign: 'center' },
					],
				],
				vertical: [
					[
						{ key: 'row-1', label: 'Row 1', textAlign: 'left' },
						{ key: 'row-2', label: 'Row 2', textAlign: 'center' },
					],
				],
			},
			_data: [
				{ header1: 'Cell 1.1', header2: 'Cell 1.2' },
				{ header1: 'Cell 2.1', header2: 'Cell 2.2' },
			],
		},
		{
			_label: 'Table with two spanned horizontal and vertical headers',
			_headerCells: {
				horizontal: [
					[{ label: 'H-Header', colSpan: 2 }],
					[
						{ key: 'header1', label: 'Sub H-Header 1' },
						{ key: 'header2', label: 'Sub H-Header 2' },
					],
				],
				vertical: [[{ label: 'V-Header', rowSpan: 2 }], [{ label: 'Sub V-Header 1' }, { label: 'Sub V-Header 2' }]],
			},
			_data: [
				{ header1: 'Cell 1.1', header2: 'Cell 1.2' },
				{ header1: 'Cell 2.1', header2: 'Cell 2.2' },
			],
		},
	],
);
