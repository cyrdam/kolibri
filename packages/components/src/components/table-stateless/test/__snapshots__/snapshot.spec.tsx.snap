// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`kol-table-stateless-wc should render with _label="Table with horizontal and vertical headers" _headerCells={"horizontal":[[{"key":"header1","label":"Header1","textAlign":"left"},{"key":"header2","label":"Header2","textAlign":"center"},{"key":"header3","label":"Header3","textAlign":"right"}]],"vertical":[[{"key":"row1","label":"Row 1","textAlign":"left"},{"key":"row2","label":"Row 2","textAlign":"center"},{"key":"row3","label":"Row 3","textAlign":"right"}]]} _data=[{"header1":"Cell 1.1","header2":"Cell 1.2","header3":"Cell 1.3"},{"header1":"Cell 2.1","header2":"Cell 2.2","header3":"Cell 2.3"},{"header1":"Cell 3.1","header2":"Cell 3.2","header3":"Cell 3.3"}] 1`] = `
<kol-table-stateless-wc class="kol-table-stateless-wc">
  <div class="table">
    <table>
      <div aria-describedby="caption" class="focus-element"></div>
      <caption id="caption">
        Table with horizontal and vertical headers
      </caption>
      <thead>
        <tr>
          <td colspan="1" rowspan="1"></td>
          <th class="align-left" data-sort="sort-undefined" scope="col">
            Header1
          </th>
          <th class="align-center" data-sort="sort-undefined" scope="col">
            Header2
          </th>
          <th class="align-right" data-sort="sort-undefined" scope="col">
            Header3
          </th>
        </tr>
        <tr aria-hidden="true" class="head-spacer">
          <td colspan="4"></td>
        </tr>
      </thead>
      <tbody>
        <tr>
          <th class="align-left" data-sort="sort-undefined" scope="row">
            Row 1
          </th>
          <td class="left" style="text-align: left;">
            Cell 1.1
          </td>
          <td class="center" style="text-align: center;">
            Cell 1.2
          </td>
          <td class="right" style="text-align: right;">
            Cell 1.3
          </td>
        </tr>
        <tr>
          <th class="align-center" data-sort="sort-undefined" scope="row">
            Row 2
          </th>
          <td class="left" style="text-align: left;">
            Cell 2.1
          </td>
          <td class="center" style="text-align: center;">
            Cell 2.2
          </td>
          <td class="right" style="text-align: right;">
            Cell 2.3
          </td>
        </tr>
        <tr>
          <th class="align-right" data-sort="sort-undefined" scope="row">
            Row 3
          </th>
          <td class="left" style="text-align: left;">
            Cell 3.1
          </td>
          <td class="center" style="text-align: center;">
            Cell 3.2
          </td>
          <td class="right" style="text-align: right;">
            Cell 3.3
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</kol-table-stateless-wc>
`;

exports[`kol-table-stateless-wc should render with _label="Table with only horizontal headers" _headerCells={"horizontal":[[{"key":"header1","label":"Header 1","textAlign":"left"},{"key":"header2","label":"Header 2","textAlign":"center"}]],"vertical":[]} _data=[{"header1":"Cell 1.1","header2":"Cell 1.2"},{"header1":"Cell 2.1","header2":"Cell 2.2"}] 1`] = `
<kol-table-stateless-wc class="kol-table-stateless-wc">
  <div class="table">
    <table>
      <div aria-describedby="caption" class="focus-element"></div>
      <caption id="caption">
        Table with only horizontal headers
      </caption>
      <thead>
        <tr>
          <th class="align-left" data-sort="sort-undefined" scope="col">
            Header 1
          </th>
          <th class="align-center" data-sort="sort-undefined" scope="col">
            Header 2
          </th>
        </tr>
        <tr aria-hidden="true" class="head-spacer">
          <td colspan="2"></td>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td class="left" style="text-align: left;">
            Cell 1.1
          </td>
          <td class="center" style="text-align: center;">
            Cell 1.2
          </td>
        </tr>
        <tr>
          <td class="left" style="text-align: left;">
            Cell 2.1
          </td>
          <td class="center" style="text-align: center;">
            Cell 2.2
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</kol-table-stateless-wc>
`;

exports[`kol-table-stateless-wc should render with _label="Table with two horizontal header rows" _headerCells={"horizontal":[[{"label":"Header 1","textAlign":"left"},{"label":"Header 2","textAlign":"center"}],[{"key":"header1","label":"Sub Header 1","textAlign":"left"},{"key":"header2","label":"Sub Header 2","textAlign":"center"}]],"vertical":[[{"key":"row-1","label":"Row 1","textAlign":"left"},{"key":"row-2","label":"Row 2","textAlign":"center"}]]} _data=[{"header1":"Cell 1.1","header2":"Cell 1.2"},{"header1":"Cell 2.1","header2":"Cell 2.2"}] 1`] = `
<kol-table-stateless-wc class="kol-table-stateless-wc">
  <div class="table">
    <table>
      <div aria-describedby="caption" class="focus-element"></div>
      <caption id="caption">
        Table with two horizontal header rows
      </caption>
      <thead>
        <tr>
          <td colspan="1" rowspan="2"></td>
          <th class="align-left" data-sort="sort-undefined" scope="col">
            Header 1
          </th>
          <th class="align-center" data-sort="sort-undefined" scope="col">
            Header 2
          </th>
        </tr>
        <tr>
          <th class="align-left" data-sort="sort-undefined" scope="col">
            Sub Header 1
          </th>
          <th class="align-center" data-sort="sort-undefined" scope="col">
            Sub Header 2
          </th>
        </tr>
        <tr aria-hidden="true" class="head-spacer">
          <td colspan="3"></td>
        </tr>
      </thead>
      <tbody>
        <tr>
          <th class="align-left" data-sort="sort-undefined" scope="row">
            Row 1
          </th>
          <td class="left" style="text-align: left;">
            Cell 1.1
          </td>
          <td class="center" style="text-align: center;">
            Cell 1.2
          </td>
        </tr>
        <tr>
          <th class="align-center" data-sort="sort-undefined" scope="row">
            Row 2
          </th>
          <td class="left" style="text-align: left;">
            Cell 2.1
          </td>
          <td class="center" style="text-align: center;">
            Cell 2.2
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</kol-table-stateless-wc>
`;

exports[`kol-table-stateless-wc should render with _label="Table with two spanned horizontal and vertical headers" _headerCells={"horizontal":[[{"label":"H-Header","colSpan":2}],[{"key":"header1","label":"Sub H-Header 1"},{"key":"header2","label":"Sub H-Header 2"}]],"vertical":[[{"label":"V-Header","rowSpan":2}],[{"label":"Sub V-Header 1"},{"label":"Sub V-Header 2"}]]} _data=[{"header1":"Cell 1.1","header2":"Cell 1.2"},{"header1":"Cell 2.1","header2":"Cell 2.2"}] 1`] = `
<kol-table-stateless-wc class="kol-table-stateless-wc">
  <div class="table">
    <table>
      <div aria-describedby="caption" class="focus-element"></div>
      <caption id="caption">
        Table with two spanned horizontal and vertical headers
      </caption>
      <thead>
        <tr>
          <td colspan="2" rowspan="2"></td>
          <th colspan="2" data-sort="sort-undefined" scope="colgroup">
            H-Header
          </th>
        </tr>
        <tr>
          <th data-sort="sort-undefined" scope="col">
            Sub H-Header 1
          </th>
          <th data-sort="sort-undefined" scope="col">
            Sub H-Header 2
          </th>
        </tr>
        <tr aria-hidden="true" class="head-spacer">
          <td colspan="4"></td>
        </tr>
      </thead>
      <tbody>
        <tr>
          <th data-sort="sort-undefined" rowspan="2" scope="row">
            V-Header
          </th>
          <th data-sort="sort-undefined" scope="row">
            Sub V-Header 1
          </th>
          <td>
            Cell 1.1
          </td>
          <td>
            Cell 1.2
          </td>
        </tr>
        <tr>
          <th data-sort="sort-undefined" scope="row">
            Sub V-Header 2
          </th>
          <td>
            Cell 2.1
          </td>
          <td>
            Cell 2.2
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</kol-table-stateless-wc>
`;
